iis-wf/iis-wf-citationmatching/src/main/java/eu/dnetlib/iis/wf/citationmatching/output/CitationMatchingOutputTransformerJob.java

```diff
@@ -33,30 +50,78 @@ public class CitationMatchingOutputTransformerJob {
+            Configuration hadoopConf = sc.hadoopConfiguration();
+            
+        	HdfsUtils.remove(hadoopConf, params.output);
+        	
+        	int numberOfEmittedFiles = params.numberOfEmittedFiles;
+            
+            LockManager lockManager = LockManagerUtils.instantiateLockManager(params.lockManagerFactoryClassName,
+                    hadoopConf);
         	
-            JavaRDD<Citation> inputCitations = avroLoader.loadJavaRDD(sc, params.input, Citation.class);
+            JavaRDD<Citation> inputMatchedCitations = avroLoader.loadJavaRDD(sc, params.inputMatchedCitations, Citation.class);
+            JavaRDD<eu.dnetlib.iis.common.citations.schemas.Citation> transformedMatchedCitations = 
+                    inputMatchedCitations.map(inputCitation -> citationToCommonCitationConverter.convert(inputCitation));
+            // caching because this RDD is used for writing in cache and at output
+            transformedMatchedCitations.cache();
+            
+            JavaPairRDD<CharSequence, CitationEntry> transformedMatchedCitationsPair = transformedMatchedCitations.mapToPair(
+                    x -> new Tuple2<CharSequence, CitationEntry>(x.getSourceDocumentId(), x.getEntry())); 
+            
+            final Path cacheRootDir = new Path(params.cacheRootDir);
+            CacheMetadataManagingProcess cacheManager = new CacheMetadataManagingProcess() ;
+            String existingCacheId = cacheManager.getExistingCacheId(hadoopConf, cacheRootDir);
             
+            // reading all cached entries
+            JavaRDD<Citations> allCachedCitations = CacheStorageUtils.getRddOrEmpty(sc, avroLoader, cacheRootDir,
+                    existingCacheId, CacheRecordType.data, Citations.class);
+            allCachedCitations.cache();
             
-            JavaRDD<eu.dnetlib.iis.common.citations.schemas.Citation> outputCitations = 
-                    inputCitations.map(inputCitation -> citationToCommonCitationConverter.convert(inputCitation));
+            JavaRDD<ExtractedDocumentMetadataMergedWithOriginal> inputDocuments = avroLoader.loadJavaRDD(sc,
+                    params.inputMetadata, ExtractedDocumentMetadataMergedWithOriginal.class);
+            JavaPairRDD<CharSequence, Integer> inputDocumentsIdtoYear = inputDocuments.mapToPair(x -> new Tuple2<CharSequence, Integer>(x.getId(), x.getYear())); 
+            // caching because this RDD is used for filtering data for both caching and generating final output
+            inputDocumentsIdtoYear.cache();
+
+            // caching only a subset of matched citations coming from a paper published earlier that X years ago
+            // and only the ones which were not cached before (required to cache also entries with no matches which will
+            // not be a part of transformedMatchedCitations but we should avoid incuding them in a subsequent run)
+            int publicationYearThreadshold = Year.now().getValue() - params.cacheOlderThanXYears;
```

Comment:
This requires use of absurd future year values in testing.
But I'm not sure if passing the current year from outside just for testing makes sense, maybe this is good enough.